lib/Transform/Clang/LoopSwapping.cpp

```diff
@@ -0,0 +1,546 @@
+    Passes.add(new ClangLoopSwappingServer);
+    Passes.add(createAnalysisWaitServerPass());
+  }
+
+  void addAfterPass(legacy::PassManager &Passes) const override {
+    Passes.add(createAnalysisReleaseServerPass());
+    Passes.add(createAnalysisCloseConnectionPass());
+  }
+};
+
+}
+
+void ClangLoopSwapping::initializeProviderOnServer() {
+  ClangLoopSwappingServerProvider::initialize<GlobalOptionsImmutableWrapper>(
+      [this](GlobalOptionsImmutableWrapper &Wrapper) {
+        Wrapper.setOptions(mGlobalOpts);
+      });
+  auto R = mSocket->getAnalysis<
+      DIMemoryEnvironmentWrapper, DIMemoryTraitPoolWrapper>();
+  assert(R && "Immutable passes must be available on server!");
+  auto *DIMEnvServer = R->value<DIMemoryEnvironmentWrapper *>();
+  ClangLoopSwappingServerProvider::initialize<DIMemoryEnvironmentWrapper>(
+      [DIMEnvServer](DIMemoryEnvironmentWrapper &Wrapper) {
+        Wrapper.set(**DIMEnvServer);
+      });
+  auto *DIMTraitPoolServer = R->value<DIMemoryTraitPoolWrapper *>();
+  ClangLoopSwappingServerProvider::initialize<DIMemoryTraitPoolWrapper>(
+      [DIMTraitPoolServer](DIMemoryTraitPoolWrapper &Wrapper) {
+        Wrapper.set(**DIMTraitPoolServer);
+      });
+      
+}
+
+std::vector<const DIAliasTrait *> ClangLoopSwapping::GetLoopTraits(MDNode *LoopID) {
+  auto DepItr = DIDepInfo->find(LoopID);
+  assert(DepItr != DIDepInfo->end() && "Loop must be analyzed!");
+  auto &DIDepSet = DepItr->get<DIDependenceSet>();
+  DenseSet<const DIAliasNode *> Coverage;
+  accessCoverage<bcl::SimpleInserter>(DIDepSet, *DIAT, Coverage,
+                                      mGlobalOpts->IgnoreRedundantMemory);
+  std::vector<const DIAliasTrait *> traits;
+  for (auto &TS : DIDepSet) {
+    auto node = TS.getNode();
+    if (!Coverage.count(node))
+      continue;
+    traits.push_back(&TS);
+  }
+  return traits;
+}
+
+bool ClangLoopSwapping::HasSameReductionKind(std::vector<const DIAliasTrait *> &traits0,
+                        std::vector<const DIAliasTrait *> &traits1,
+                        SpanningTreeRelation<DIAliasTree *> &STR) {
+  for (auto TS0: traits0) {
+    auto *node0 = const_cast<DIAliasNode *>(TS0->getNode());
+    MemoryDescriptor Dptr0 = *TS0;
+    for (auto TS1: traits1) {
+      auto *node1 = const_cast<DIAliasNode *>(TS1->getNode());
+      MemoryDescriptor Dptr1 = *TS1;
+      if (STR.isEqual(node0, node1) && Dptr0.is<trait::Reduction>() &&
+          Dptr1.is<trait::Reduction>()) {
+        dbgs() << "[LOOP SWAPPING]: same reduction nodes;\n";
+        auto I0 = TS0->begin(), I1 = TS1->begin();
+        auto *Red0 = (**I0).get<trait::Reduction>();
+        auto *Red1 = (**I1).get<trait::Reduction>();
+        auto Kind0 = Red0->getKind(), Kind1 = Red1->getKind();
+        if (!Red0 || !Red1 || Kind0 == trait::DIReduction::RK_NoReduction ||
+            Kind1 == trait::DIReduction::RK_NoReduction) {
+          dbgs() << "[LOOP SWAPPING]: No Reduction\n";
+          return true;
```

Comment:
Странно, что здесь сразу стоит `return`, ведь к этому моменту возможно еще не все редукции проверены.
```suggestion
          continue;
```